{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "black-simple",
   "metadata": {},
   "source": [
    " <font size=\"10\">**Pseudo-spatial visualisation: Plotly; outside functions**</font>\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "regulated-palestinian",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    \n",
    "<b> <h1> ℹ️ <strong> <font size=\"6\" color=\"black\"> Important notebook information </font> </strong> </h1> </b>\n",
    "    <hr>\n",
    "    <font size=\"4\" color=\"black\">\n",
    "        The purpose of this notebook is use the combined scRNAseq data and masks object generated to plot the data using masks set included <br> <br>\n",
    "    This notebook everything is run out of functions for easier manipulation</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vertical-prospect",
   "metadata": {},
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prospective-security",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.lines import Line2D\n",
    "import matplotlib.dates as mdates\n",
    "import seaborn as sns\n",
    "import cv2\n",
    "import joypy\n",
    "import math\n",
    "import scipy\n",
    "from scipy import stats\n",
    "import re\n",
    "import datetime\n",
    "from dateutil.relativedelta import relativedelta\n",
    "import plotly\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "from ipywidgets import Button, Dropdown, HBox, VBox\n",
    "from skimage.draw import polygon\n",
    "from PIL import Image\n",
    "from IPython.display import display\n",
    "from plotly.offline import iplot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gentle-hurricane",
   "metadata": {},
   "source": [
    "# Load in anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "differential-chemistry",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('anndata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "biological-recommendation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check anndata object\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pacific-blair",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check column dtypes\n",
    "adata.obs.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "marked-cabin",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check mask sets and correlated column to use \n",
    "adata.uns['Mask_selector']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "opposite-frequency",
   "metadata": {},
   "source": [
    "# User set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "downtown-touch",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to compare gene expression of 1 celltype of interest / group of interest cut beforehand and then run this \n",
    "\n",
    "# Calculate tables to hold potential data to plot\n",
    "\n",
    "cat1 = 'region_col'    # make this the column which the masks relate to i.e. 12 sections\n",
    "cat2 = 'col_interest'  # change to annotations of interest\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments Elmer:\n",
    "\n",
    "# manual = manual\n",
    "\n",
    "# celltype_counts = the total number of cells for specified celltype within a section\n",
    "\n",
    "# celltype_percentage_within_sections = the percentage of the specified celltype within a section compared against all other celltypes within the same section\n",
    "\n",
    "# celltype_percentage_across_sections = the percentage of only the specified celltype compared across all sections i.e. the celltypes broad distribution across sections\n",
    "\n",
    "# gene_expression = average gene expression of a specified gene for a section using all cells within the section. Note: if ran sc.pp.scale then this is average z-scores of the expression values otherwise it is just average expression values\n",
    "\n",
    "# proportion = bool\n",
    "\n",
    "# Mandatory arguments\n",
    "mask_set = 'set_1'\n",
    "\n",
    "mode = 'mode_interest'   # manual, celltype_counts, celltype_percentage_within_sections, celltype_percentage_across_sections, gene_expression\n",
    "\n",
    "# Input celltype of gene of interest\n",
    "plot_value = ['value']  #['value_1','value_2','value_3']\n",
    "#plot_value = list(adata.obs['anno_col'].unique())\n",
    "#plot_value = [True]\n",
    "\n",
    "scale = 'auto'             # for the color bar: auto, manual\n",
    "cmap = plt.cm.viridis      # using premade colormaps e.g. viridis, plasma, inferno, magma, cividis, Reds\n",
    "\n",
    "# if chose scale manual\n",
    "scale_lower_value = 0\n",
    "scale_upper_value = 20\n",
    "\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments continuous and int variables:\n",
    "\n",
    "# log the column of interest \n",
    "scale_log = False\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments for datetime columns\n",
    "\n",
    "plot_covid = False\n",
    "use_premade_info = True\n",
    "cat3 = 'ID_col'\n",
    "\n",
    "#################################################################################\n",
    "\n",
    "# If chose manual mode - input manual values below\n",
    "manual_values = [\n",
    "42,   # Section 1\n",
    "3,   # Section 2\n",
    "31,   # Section 3\n",
    "41,   # Section 4\n",
    "15,   # Section 5\n",
    "26,   # Section 6\n",
    "67,   # Section 7\n",
    "28,   # Section 8\n",
    "91,   # Section 9\n",
    "10,   # Section 10\n",
    "13,   # Section 11\n",
    "22,   # Section 12\n",
    "]\n",
    "\n",
    "save_plot = './plot_X.pdf' # False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "frequent-request",
   "metadata": {},
   "source": [
    "# Generate plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "corrected-shelf",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################################\n",
    "if adata.obs[cat2].dtype == \"bool\":\n",
    "        adata.obs[cat2] = adata.obs[cat2].astype(\"str\").astype(\"category\")\n",
    "\n",
    "\n",
    "\n",
    "if pd.api.types.is_categorical_dtype(adata.obs[cat2]):\n",
    "    col_date_test = []\n",
    "    for i in adata.obs[cat2].unique():\n",
    "        if re.search(\"^(\\d{4})-(0[1-9]|1[0-2]|[1-9])-([1-9]|0[1-9]|[1-2]\\d|3[0-1])$\", i):\n",
    "            col_date_test.append(True)\n",
    "            if len(col_date_test) == len(adata.obs[cat2].unique()):\n",
    "                adata.obs[cat2] = adata.obs[cat2].astype('datetime64[ns]')\n",
    "        else:\n",
    "            break\n",
    "else:\n",
    "    pass\n",
    "######################################################################################\n",
    "\n",
    "if adata.obs[cat2].dtype == 'datetime64[ns]':\n",
    "    if use_premade_info == True:\n",
    "        Dates = adata.uns[mask_set + '_' + cat2]['dates']\n",
    "        dates = []\n",
    "        for d in Dates:\n",
    "            dates.append(datetime.date(*list(d)))\n",
    "        labels = ['{0:%d %b %Y}:<br>{1}'.format(d, l) for l, d in zip (adata.uns[mask_set + '_' + cat2]['labels'], dates)]\n",
    "        #dates = adata.uns[cat2]['dates']\n",
    "    else:\n",
    "        #cat3 = 'Embryo_no' \n",
    "        df_new = ((adata.obs[[cat3,cat2]]).reset_index(drop=True).drop_duplicates().reset_index(drop=True)).set_index(cat3)\n",
    "        date_dict = {}\n",
    "        for i in df_new.index:\n",
    "            date_dict[i] = df_new.loc[i][0]\n",
    "        labels = ['{0:%d %b %Y}:<br>{1}'.format(d, l) for l, d in zip (date_dict.keys(), date_dict.values())]\n",
    "        dates = date_dict.values()\n",
    "        dates = [i.date() for i in dates]\n",
    "\n",
    "\n",
    "    if plot_covid == True:\n",
    "        covid_start = datetime.date(2020, 3, 23)\n",
    "        covid_end = datetime.date(2022, 2, 24)\n",
    "    \n",
    "        if covid_start < min(dates):\n",
    "            start_date = (covid_start - relativedelta(months=1)).replace(day=1)\n",
    "        else:\n",
    "            start_date = (min(dates) - relativedelta(months=1)).replace(day=1)\n",
    "            \n",
    "        if covid_end > max(dates):\n",
    "            end_date = (covid_end + relativedelta(months=1)).replace(day=1)\n",
    "        else:\n",
    "            end_date = (max(dates) + relativedelta(months=1)).replace(day=1)\n",
    "    \n",
    "    else:\n",
    "        start_date = (min(dates) - relativedelta(months=1)).replace(day=1)\n",
    "        end_date = (max(dates) + relativedelta(months=1)).replace(day=1)\n",
    "    \n",
    "    \n",
    "    stems = np.zeros(len(dates))\n",
    "    stems[::2] = 1#0.3\n",
    "    stems[1::2] = -1# -0.3\n",
    "\n",
    "    data = [\n",
    "        go.Scatter(\n",
    "            x = dates, \n",
    "            #y = np.zeros(len(dates)),\n",
    "            y=stems,\n",
    "            mode='markers',\n",
    "            marker=dict(color='red'),\n",
    "            text=labels,\n",
    "            hoverinfo = 'text+x+y',\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    # Use the 'shapes' attribute from the layout to draw the vertical lines\n",
    "    layout = go.Layout(\n",
    "        shapes=[dict(\n",
    "            type='line',\n",
    "            xref='x',\n",
    "            yref='y',\n",
    "            x0=i,\n",
    "            y0=0,\n",
    "            x1=i,\n",
    "            y1=stems[dates.index(i)],\n",
    "            line=dict(\n",
    "                color='black',\n",
    "                width=1\n",
    "            )\n",
    "        ) for i in dates],\n",
    "        title=f'Timeline for {cat2}'\n",
    "    )\n",
    "\n",
    "    # Plot the chart\n",
    "    fig = go.Figure(data, layout)\n",
    "\n",
    "    fig.add_hline(y=0)\n",
    "\n",
    "    # add month ticks and labels \n",
    "    for i in list((pd.date_range(start=start_date, end=end_date, freq='1MS')).map(lambda d: str(d.date()))):\n",
    "        x_axis_pos = datetime.date(*list(map(int, i.replace('-', ' ').split(' '))))\n",
    "        fig.add_vline(x=x_axis_pos, y0=0.45,y1=0.55, line_width=1, line_color=\"black\") \n",
    "        if (x_axis_pos.strftime(\"%B\")[:3]) == 'Jan':\n",
    "            fig.add_annotation(dict(font=dict(color='black',size=7),\n",
    "                                            x=x_axis_pos,\n",
    "                                            y=-0.3,\n",
    "                                            showarrow=False,\n",
    "                                            text=x_axis_pos.strftime(\"%B\")[:3],\n",
    "                                            textangle=0,\n",
    "                                            xanchor='left',\n",
    "                                            xref=\"x\",\n",
    "                                            yref=\"y\"))\n",
    "            fig.add_annotation(dict(font=dict(color='black',size=10),\n",
    "                                            x=x_axis_pos,\n",
    "                                            y=0.5,\n",
    "                                            showarrow=False,\n",
    "                                            text=x_axis_pos.strftime(\"%Y\"),\n",
    "                                            textangle=0,\n",
    "                                            xanchor='left',\n",
    "                                            xref=\"x\",\n",
    "                                            yref=\"y\"))\n",
    "        else:\n",
    "            fig.add_annotation(dict(font=dict(color='black',size=7),\n",
    "                                            x=x_axis_pos,\n",
    "                                            y=-0.3,\n",
    "                                            showarrow=False,\n",
    "                                            text=x_axis_pos.strftime(\"%B\")[:3],\n",
    "                                            textangle=0,\n",
    "                                            xanchor='left',\n",
    "                                            xref=\"x\",\n",
    "                                            yref=\"y\"))\n",
    "\n",
    "    if plot_covid == True:\n",
    "        fig.add_vline(x=covid_start, y0=0.55,y1=0.65, line_width=1, line_color=\"purple\")\n",
    "        fig.add_vline(x=covid_end, y0=0.55,y1=0.65, line_width=1, line_color=\"purple\")\n",
    "        fig.add_shape(type='line',\n",
    "                        x0=covid_start,\n",
    "                        y0=0.25,\n",
    "                        x1=covid_end,\n",
    "                        y1=0.25,\n",
    "                        line=dict(color='purple',width=1),\n",
    "                        xref='x',\n",
    "                        yref='y'\n",
    "        )\n",
    "        fig.add_annotation(dict(font=dict(color='purple',size=10),\n",
    "                                            x=covid_start,\n",
    "                                            y=0.5,\n",
    "                                            showarrow=False,\n",
    "                                            text='Covid',\n",
    "                                            textangle=0,\n",
    "                                            xanchor='left',\n",
    "                                            xref=\"x\",\n",
    "                                            yref=\"y\"))\n",
    "\n",
    "    \n",
    "    \n",
    "    fig.update_xaxes(visible=False, fixedrange=False, autorange=True,  rangeslider=dict(autorange=True,thickness=0.3, bgcolor=\"#e4f7fe\")) #bgcolor=\"#ADD8E6\" #range=[start_date, end_date],\n",
    "    fig.update_yaxes(visible=False, fixedrange=True)\n",
    "    fig.update_layout({'plot_bgcolor': 'rgba(0,0,0,0)','paper_bgcolor': 'rgba(0,0,0,0)'}, autosize=False, width=600, height=400, )\n",
    "\n",
    "    fig.update_xaxes(type=\"date\", dtick=\"M1\")\n",
    "\n",
    "    #fig.show()\n",
    "    \n",
    "\n",
    "\n",
    "######################################################################################\n",
    "    \n",
    "elif adata.obs[cat2].dtype in [\"float64\", \"int32\", \"int64\"]:\n",
    "    fig = go.Figure()\n",
    "    for index, section in enumerate(adata.obs[cat1].unique()):\n",
    "        values = (adata.obs[cat2][adata.obs[cat1].isin([section])]).values\n",
    "        \n",
    "        if scale_log == True:\n",
    "            values = np.log(values)\n",
    "        \n",
    "        l = len(adata.obs[cat1].unique())\n",
    "        c = cm.get_cmap(cmap, l)\n",
    "        fig.add_trace(go.Violin(x=values, line_color=f'rgb{c(index/l)[:3]}', name=section))\n",
    "\n",
    "    fig.update_traces(orientation='h', side='positive', width=2, points=False)\n",
    "    fig.update_layout(xaxis_showgrid=False, xaxis_zeroline=False, \n",
    "                      title_text=f'Ridgeplot of the continual variable {cat2} across {cat1}', \n",
    "                      legend_title=\"Trace selector\",\n",
    "                     xaxis_title=f\"{cat2}\",\n",
    "       yaxis_title=f\"{cat1}\",)\n",
    "    fig.update_layout({'plot_bgcolor': 'rgba(0,0,0,0)','paper_bgcolor': 'rgba(0,0,0,0)'}, autosize=False, width=800, height=800, )\n",
    "\n",
    "    #fig.show()\n",
    "    \n",
    "    \n",
    "######################################################################################\n",
    "\n",
    "    \n",
    "else:\n",
    "    if mode == 'manual':\n",
    "        values = manual_values\n",
    "    elif mode == 'gene_expression':\n",
    "        df_of_values = (adata.varm[mask_set + '_Sectional_gene_expression'].T)[plot_value]\n",
    "        df_of_values = df_of_values.T\n",
    "        values = []\n",
    "        for col in df_of_values:\n",
    "            value = list(df_of_values[col].values)\n",
    "            values.extend(value)\n",
    "\n",
    "    elif mode in ['celltype_counts','celltype_percentage_within_sections','celltype_percentage_across_sections', 'proportion']:\n",
    "\n",
    "        if len(plot_value) > 1:\n",
    "            adata.obs['combined_annotation'] = adata.obs[cat2].copy().astype(str)\n",
    "            for value in plot_value:\n",
    "                adata.obs.loc[adata.obs[cat2].isin([value]), 'combined_annotation'] = 'combined_annotation'\n",
    "            cat2 = 'combined_annotation'\n",
    "            plot_value = ['combined_annotation']\n",
    "\n",
    "        adata.obs[cat1] = adata.obs[cat1].astype('category')\n",
    "        adata.obs[cat2] = adata.obs[cat2].astype('category')\n",
    "\n",
    "        # generate counts table\n",
    "        counts_table = pd.crosstab(adata.obs[cat1],adata.obs[cat2])\n",
    "\n",
    "        if mode == 'celltype_counts':\n",
    "            df_of_values = counts_table[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "\n",
    "        elif mode == 'celltype_percentage_across_sections':\n",
    "            percentage_table_column = (round((counts_table/counts_table.sum())*100,2))\n",
    "            df_of_values = percentage_table_column[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "\n",
    "\n",
    "        elif mode == 'celltype_percentage_within_sections':\n",
    "            percentage_table_row = (round(((counts_table.T)/(counts_table.T).sum())*100,2)).T\n",
    "            df_of_values = percentage_table_row[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "        \n",
    "        elif mode == 'proportion':\n",
    "            values = (counts_table[\"True\"]/counts_table.sum(axis=1)*100).values\n",
    "\n",
    "    else:\n",
    "        raise Exception('Mode option not correct. Please use one of the following: manual, celltype_counts, celltype_percentage, gene_expression or proportion')\n",
    "\n",
    "    ##################################################################################\n",
    "\n",
    "    # create a color scale on the range of values inputted\n",
    "    \n",
    "    values_dict = dict(zip(adata.obs[cat1].unique(),values))\n",
    "    \n",
    "    if scale == 'auto':\n",
    "        norm = mpl.colors.Normalize( vmin=min(values) , vmax=max(values) )\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "\n",
    "    elif scale == 'manual':\n",
    "        norm = mpl.colors.Normalize( vmin=scale_lower_value , vmax=scale_upper_value )\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    else:\n",
    "        raise Exception('Scale option not correct. Please use either auto or manual')    \n",
    "    \n",
    "    \n",
    "    ##################################################################################\n",
    "    \n",
    "    fig = go.Figure()\n",
    "                        \n",
    "    count=0\n",
    "    for key in adata.uns[mask_set+'_polygons'].keys():\n",
    "        if len(plot_value) == len(adata.obs[cat2].unique()):\n",
    "            title = 'All options in category selected'\n",
    "        else:\n",
    "            colorbar_trace=go.Scatter(x=[None],\n",
    "                         y=[None],\n",
    "                         mode='markers',\n",
    "                         marker=dict(\n",
    "                             colorscale='viridis', \n",
    "                             showscale=True,\n",
    "                             cmin=min(values),\n",
    "                             cmax=max(values),\n",
    "                             colorbar=dict(thickness=10, tickmode='auto'), #tickvals=[min(values), max(values)],\n",
    "                         ),\n",
    "                         hoverinfo='none',\n",
    "                        showlegend=False\n",
    "                        )\n",
    "            fig.add_trace(colorbar_trace)\n",
    "\n",
    "            if mode == 'manual':\n",
    "                title = 'Manual values for every section'\n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"Manual value inputted: {plot_value[0]}\"])\n",
    "            elif mode == 'gene_expression':\n",
    "                title = f'Mean gene expression of {plot_value[0]} for each section'\n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"Section mean gene expression value: {plot_value[0]}\"])\n",
    "            elif mode == 'celltype_counts':\n",
    "                title = f'Number of counts for {plot_value[0]}'                \n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"Number of {plot_value[0]} cells in {key}:    {[val for k, val in values_dict.items() if k in key][0]} cells\",\n",
    "                        f\"Total number of {plot_value[0]} cells in data:    {sum(values)} cells\"])\n",
    "            elif mode == 'celltype_percentage_within_sections':\n",
    "                title = f'Percentage of {plot_value[0]} compared within section'\n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"{plot_value[0]} represents {values[count]}% of the cells within {key}\"])\n",
    "            elif mode == 'celltype_percentage_across_sections':\n",
    "                title = f'Percentage of {plot_value[0]} across sections'\n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"{key} contains {values[count]}% of the cells for {plot_value[0]}\"])\n",
    "            elif mode == 'proportion':\n",
    "                title = f'Percentage of truthful {cat2} values within section'\n",
    "                text = \"<br>\".join([\n",
    "                        f\"<b>{key}</b>\",\n",
    "                        \"\",\n",
    "                        f\"{round(values[count],2)}% (2dp) of the cells within {key} are true for {cat2}\"])\n",
    "                                       \n",
    "            \n",
    "        polygon0 = go.Scatter(\n",
    "            x=list(*adata.uns[mask_set+'_polygons'][key][:, :, 0, 0]),\n",
    "            y=list(*adata.uns[mask_set+'_polygons'][key][:, :, 0, 1]),\n",
    "            showlegend=False,\n",
    "            mode=\"lines\",\n",
    "            fill='toself',\n",
    "            line=dict(color=(\"#%02x%02x%02x\" % tuple(list(int((255 * x))for x in list(sm.to_rgba([val for k, val in values_dict.items() if k in key][0]))[0:3]))),width=1,),\n",
    "            fillcolor=(\"#%02x%02x%02x\" % tuple(list(int((255 * x))for x in list(sm.to_rgba([val for k, val in values_dict.items() if k in key][0]))[0:3]))),\n",
    "            hoveron = 'fills',\n",
    "            text=text,\n",
    "            hoverinfo = 'text+x+y',\n",
    "            #title = title\n",
    "            )\n",
    "        fig.add_trace(polygon0)\n",
    "        fig.update_layout(title_text=title)\n",
    "        count+=1     \n",
    "    \n",
    "    fig.update_xaxes(visible=False, fixedrange=True)\n",
    "    fig.update_yaxes(visible=False, autorange=\"reversed\", fixedrange=True)\n",
    "    fig.update_layout({'plot_bgcolor': 'rgba(0,0,0,0)','paper_bgcolor': 'rgba(0,0,0,0)'}, autosize=False, width=600, height=800, )\n",
    "\n",
    "\n",
    "fig.show()\n",
    "\n",
    "if save_plot != False:\n",
    "    fig.write_image(save_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "atmospheric-asbestos",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "spectacular-amino",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rising-mozambique",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python_env_Hlab",
   "language": "python",
   "name": "python_env_hlab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
