{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "continuing-bangladesh",
   "metadata": {},
   "source": [
    " <font size=\"10\">**Pseudo-spatial visualisation: Matplotlib; outside functions**</font>\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "french-lottery",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    \n",
    "<b> <h1> ℹ️ <strong> <font size=\"6\" color=\"black\"> Important notebook information </font> </strong> </h1> </b>\n",
    "    <hr>\n",
    "    <font size=\"4\" color=\"black\">\n",
    "        The purpose of this notebook is use the combined scRNAseq data and masks object generated to plot the data using masks set included <br> <br>\n",
    "    This notebook everything is run out of functions for easier manipulation</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "broken-framing",
   "metadata": {},
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "happy-recycling",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.lines import Line2D\n",
    "import matplotlib.dates as mdates\n",
    "import seaborn as sns\n",
    "import cv2\n",
    "import joypy\n",
    "import math\n",
    "import scipy\n",
    "from scipy import stats\n",
    "import re\n",
    "import datetime\n",
    "from dateutil.relativedelta import relativedelta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "consistent-designer",
   "metadata": {},
   "source": [
    "# Load in anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "conventional-credits",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('anndata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "preceding-crest",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check data\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "through-stockholm",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check column dtypes\n",
    "adata.obs.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "entertaining-conclusion",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for gene presence if required\n",
    "adata.var[adata.var_names.str.startswith('HOX')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "funded-import",
   "metadata": {},
   "source": [
    "# User set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "normal-absolute",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to compare gene expression of 1 celltype of interest / group of interest cut beforehand and then run this \n",
    "\n",
    "# Calculate tables to hold potential data to plot\n",
    "\n",
    "cat1 = 'spatial_location'    # make this the column which the masks relate to i.e. 12 sections\n",
    "cat2 = 'anno_col'  # change to annotations of interest\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments Elmer:\n",
    "\n",
    "# manual = manual\n",
    "\n",
    "# celltype_counts = the total number of cells for specified celltype within a section\n",
    "\n",
    "# celltype_percentage_within_sections = the percentage of the specified celltype within a section compared against all other celltypes within the same section\n",
    "\n",
    "# celltype_percentage_across_sections = the percentage of only the specified celltype compared across all sections i.e. the celltypes broad distribution across sections\n",
    "\n",
    "# gene_expression = average gene expression of a specified gene for a section using all cells within the section. Note: if ran sc.pp.scale then this is average z-scores of the expression values otherwise it is just average expression values\n",
    "\n",
    "\n",
    "# Mandatory arguments\n",
    "mask_set = 'set1'\n",
    "\n",
    "mode = 'gene_expression'   # manual, celltype_counts, celltype_percentage_within_sections, celltype_percentage_across_sections, gene_expression\n",
    "\n",
    "# Input celltype of gene of interest\n",
    "plot_value = ['HOXD-AS2']  #['celltype_1','celltype_2','celltype_3']\n",
    "#plot_value = list(adata.obs['anno_col'].unique())\n",
    "\n",
    "scale = 'auto'             # for the color bar: auto, manual\n",
    "cmap = plt.cm.viridis      # using premade colormaps e.g. viridis, plasma, inferno, magma, cividis, Reds\n",
    "\n",
    "# if chose scale manual\n",
    "scale_lower_value = 0\n",
    "scale_upper_value = 20\n",
    "\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments continuous and int variables:\n",
    "\n",
    "# number of points between the upper and lower plot values - won't work for logged scale\n",
    "tick_no = 8\n",
    "\n",
    "# log the column of interest \n",
    "scale_log = False\n",
    "\n",
    "#################################################################################\n",
    "# Mode arguments for datetime columns\n",
    "\n",
    "plot_covid = True\n",
    "use_premade_info = True\n",
    "cat3 = 'ID_col'\n",
    "\n",
    "#################################################################################\n",
    "\n",
    "# If chose manual mode - input manual values below\n",
    "manual_values = [\n",
    "42,   # Section 1\n",
    "3,   # Section 2\n",
    "31,   # Section 3\n",
    "41,   # Section 4\n",
    "15,   # Section 5\n",
    "26,   # Section 6\n",
    "67,   # Section 7\n",
    "28,   # Section 8\n",
    "91,   # Section 9\n",
    "10,   # Section 10\n",
    "13,   # Section 11\n",
    "22,   # Section 12\n",
    "]\n",
    "\n",
    "save_plot = './plot_X.pdf' # False\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adverse-lindsay",
   "metadata": {},
   "source": [
    "# Generate plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dense-uncertainty",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################################\n",
    "# column checker - check if column can be read as datetime:\n",
    "#if adata.obs[cat2].dtype == 'category':\n",
    "if pd.api.types.is_categorical_dtype(adata.obs[cat2]):\n",
    "    col_date_test = []\n",
    "    for i in adata.obs[cat2].unique():\n",
    "        if re.search(\"^(\\d{4})-(0[1-9]|1[0-2]|[1-9])-([1-9]|0[1-9]|[1-2]\\d|3[0-1])$\", i):\n",
    "            col_date_test.append(True)\n",
    "            if len(col_date_test) == len(adata.obs[cat2].unique()):\n",
    "                adata.obs[cat2] = adata.obs[cat2].astype('datetime64[ns]')\n",
    "        else:\n",
    "            break\n",
    "else:\n",
    "    pass\n",
    "######################################################################################\n",
    "\n",
    "if adata.obs[cat2].dtype == 'datetime64[ns]':\n",
    "    if use_premade_info == True:\n",
    "        Dates = adata.uns[cat2]['dates']\n",
    "        dates = []\n",
    "        for d in Dates:\n",
    "            dates.append(datetime.date(*list(d)))\n",
    "        labels = ['{0:%d %b %Y}:\\n{1}'.format(d, l) for l, d in zip (adata.uns[cat2]['labels'], dates)]\n",
    "        #dates = adata.uns[cat2]['dates']\n",
    "    else:\n",
    "        #cat3 = 'Embryo_no' \n",
    "        df_new = ((adata.obs[[cat3,cat2]]).reset_index(drop=True).drop_duplicates().reset_index(drop=True)).set_index(cat3)\n",
    "        date_dict = {}\n",
    "        for i in df_new.index:\n",
    "            date_dict[i] = df_new.loc[i][0]\n",
    "        labels = ['{0:%d %b %Y}:\\n{1}'.format(d, l) for l, d in zip (date_dict.keys(), date_dict.values())]\n",
    "        dates = date_dict.values()\n",
    "\n",
    "\n",
    "    if plot_covid == True:\n",
    "        covid_start = datetime.date(2020, 3, 23)\n",
    "        covid_end = datetime.date(2022, 2, 24)\n",
    "    \n",
    "        if covid_start < min(dates):\n",
    "            start_date = (covid_start - relativedelta(months=1)).replace(day=1)\n",
    "        else:\n",
    "            start_date = (min(dates) - relativedelta(months=1)).replace(day=1)\n",
    "            \n",
    "        if covid_end > max(dates):\n",
    "            end_date = (covid_end + relativedelta(months=1)).replace(day=1)\n",
    "        else:\n",
    "            end_date = (max(dates) + relativedelta(months=1)).replace(day=1)\n",
    "    \n",
    "    else:\n",
    "        start_date = (min(dates) - relativedelta(months=1)).replace(day=1)\n",
    "        end_date = (max(dates) + relativedelta(months=1)).replace(day=1)\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(10, 4), constrained_layout=True)\n",
    "    _ = ax.set_ylim(-2, 1.75)\n",
    "    _ = ax.set_xlim(start_date,end_date)\n",
    "    _ = ax.axhline(0, xmin=0, xmax=1, c='red', zorder=1)\n",
    "    \n",
    "    _ = ax.get_xaxis().set_major_locator(mdates.MonthLocator(interval=1))\n",
    "    _ = ax.get_xaxis().set_major_formatter(mdates.DateFormatter(\"%b %Y\"))\n",
    "    \n",
    "    _ = ax.scatter(dates, np.zeros(len(dates)), s=120, c='green', zorder=2)\n",
    "    _ = ax.scatter(dates, np.zeros(len(dates)), s=30, c='darkgreen', zorder=3)\n",
    "    \n",
    "    label_offsets = np.zeros(len(dates))\n",
    "    label_offsets[::2] = 1.4#0.35\n",
    "    label_offsets[1::2] = -2.3#-0.7\n",
    "    for i, (l, d) in enumerate(zip(labels, dates)):\n",
    "        _ = ax.text(d, label_offsets[i], l, ha='center', fontfamily='serif', fontweight='bold', color='royalblue',fontsize=12)\n",
    "    \n",
    "    \n",
    "    stems = np.zeros(len(dates))\n",
    "    stems[::2] = 1#0.3\n",
    "    stems[1::2] = -1# -0.3   \n",
    "    markerline, stemline, baseline = ax.stem(dates, stems, use_line_collection=True)\n",
    "    _ = plt.setp(markerline, marker=',', color='green', markersize=5)\n",
    "    _ = plt.setp(stemline, color='green', linewidth=1.25)\n",
    "    \n",
    "    # hide lines around chart\n",
    "    for spine in [\"left\", \"top\", \"right\", \"bottom\"]: \n",
    "        _ = ax.spines[spine].set_visible(False)\n",
    "        \n",
    "    \n",
    "    _ = ax.set_title(f'Timeline for {cat2}', fontweight=\"bold\", fontfamily='serif', fontsize=16, \n",
    "                     color='royalblue', y=1.4)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # add month ticks and labels \n",
    "    for i in list((pd.date_range(start=start_date, end=end_date, freq='1MS')).map(lambda d: str(d.date()))):\n",
    "        x_axis_pos = datetime.date(*list(map(int, i.replace('-', ' ').split(' '))))\n",
    "        _ = ax.axvline(x_axis_pos, ymin=0.5, ymax=0.57, c='black', zorder=1)\n",
    "        if (x_axis_pos.strftime(\"%B\")[:3]) == 'Jan':\n",
    "            _ = ax.text(x_axis_pos, -0.3, x_axis_pos.strftime(\"%B\")[:3])\n",
    "            _ = ax.text(x_axis_pos, 0.5, x_axis_pos.strftime(\"%Y\"))\n",
    "        else:\n",
    "            _ = ax.text(x_axis_pos, -0.3, x_axis_pos.strftime(\"%B\")[:3])\n",
    "    \n",
    "    if plot_covid == True:\n",
    "        # add covid line segment\n",
    "        x_min, x_max = ax.get_xlim()\n",
    "        ticks = [(tick - x_min)/(x_max - x_min) for tick in ax.get_xticks()]\n",
    "        tick_labels = ax.get_xticklabels()\n",
    "        _ = ax.axvline(covid_start, ymin=0.6, ymax=0.7, c='purple', zorder=1)\n",
    "        _ = ax.axvline(covid_end, ymin=0.6, ymax=0.7, c='purple', zorder=1)\n",
    "        _ = ax.plot([covid_start,covid_end], [0.4,0.4], linestyle='-', color='purple')\n",
    "        _ = ax.text(covid_start,0.7, \"Covid restrictions\",color='purple', fontsize=14)\n",
    "                 \n",
    "    # hide tick labels\n",
    "    _ = ax.set_xticks([])\n",
    "    _ = ax.set_yticks([])\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "######################################################################################\n",
    "\n",
    "#if (set(adata.obs[cat2].cat.categories) == set(['True','False'])) == True:\n",
    "\n",
    "elif adata.obs[cat2].dtype == bool:\n",
    "    \n",
    "    sub_df = adata.obs[[cat1, cat2]]\n",
    "    plot_df = pd.DataFrame(index=['True','False'])\n",
    "    \n",
    "    for s in sub_df[cat1].unique():\n",
    "        df_sub = sub_df[sub_df[cat1].isin([s])]\n",
    "        true_counts = df_sub[cat2].sum() #.apply(eval).astype('boolean')\n",
    "        total = len(df_sub[cat2])\n",
    "        t_percent = round((true_counts/total)*100,2)\n",
    "        f_percent = 100 - t_percent\n",
    "        plot_df[s] = [t_percent,f_percent]\n",
    "\n",
    "\n",
    "    plot_df = plot_df.T\n",
    "    plot_df = plot_df.iloc[::-1]\n",
    "    \n",
    "    ax = plot_df.plot(kind='barh', stacked=True, legend=False, figsize = (10,10), cmap= cmap)\n",
    "    \n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    \n",
    "    for p in ax.patches:\n",
    "        width, height = p.get_width(), p.get_height()\n",
    "        x, y = p.get_xy() \n",
    "        ax.text(x+width/2, \n",
    "                y+height+0.2, \n",
    "                '{:.2f}%'.format(width), \n",
    "                horizontalalignment='center', \n",
    "                verticalalignment='center',\n",
    "                color='black',\n",
    "                fontsize=10,\n",
    "                )\n",
    "    \n",
    "    for tick in ax.yaxis.get_major_ticks():\n",
    "        tick.label.set_position((-0.03, 0))\n",
    "        tick.label.set_fontsize(12)\n",
    "    \n",
    "    plt.tick_params(left = False)\n",
    "        \n",
    "    plt.xlabel('Percentage (%)', fontsize=14)\n",
    "    yl = plt.ylabel(cat1, fontsize=16, color='black', alpha=1)    \n",
    "    \n",
    "    plt.title(f'Boolean proportion percentage of {cat2} per anatomical section', fontsize=16, y=1.03)   \n",
    "    \n",
    "    plt.legend(title=f'{cat2}', bbox_to_anchor=(1.4, 1), labelspacing = 2, fontsize=12, title_fontsize=15, prop={'size': 15})\n",
    "\n",
    "\n",
    "\n",
    "######################################################################################\n",
    "\n",
    "elif adata.obs[cat2].dtype == 'float64' or adata.obs[cat2].dtype == 'int':\n",
    "    \n",
    "    if scale_log == True:\n",
    "        adata.obs['log_scaled_col'] = np.log10(adata.obs[cat2])\n",
    "        cat_use = 'log_scaled_col'\n",
    "    else:\n",
    "        cat_use = cat2\n",
    "    \n",
    "    x_lim_range = [adata.obs[[cat_use,cat1]][cat_use].min(),adata.obs[[cat_use,cat1]][cat_use].max()]\n",
    "    \n",
    "    array = []\n",
    "    ticks = np.linspace(int(x_lim_range[0]), int(x_lim_range[1]), num=8, dtype=int)\n",
    "    array.append(ticks[0])\n",
    "        \n",
    "    if scale_log == True:\n",
    "        for i in ticks[1:]:\n",
    "            array.append(math.ceil(i))\n",
    "    else:\n",
    "        for i in ticks[1:]:\n",
    "            array.append(math.ceil(i / 100) *100)\n",
    "    \n",
    "    fig, axes = joypy.joyplot(data=adata.obs[[cat_use,cat1]], by=cat1, colormap=cmap, fade = True, range_style='group', x_range=x_lim_range, tails=0, xlim='max',  figsize = (10,10), overlap = 0, ylabelsize=12, xlabelsize=12) \n",
    "    \n",
    "    plt.title(f'Ridgeplot of the continual variable {cat2} across {cat1}', fontsize=16, y=1.03)\n",
    "    axes[-1].set_xticks(array)\n",
    "    if scale_log == True:\n",
    "        axes[-1].set_xlabel(f'log10 of {cat2}', fontsize=16, color='black', alpha=1)\n",
    "    else:\n",
    "        axes[-1].set_xlabel(cat2, fontsize=16, color='black', alpha=1)\n",
    "    axes[-1].xaxis.set_label_coords(0.5,-0.07)\n",
    "    \n",
    "    axes[-1].yaxis.set_visible(True)\n",
    "    axes[-1].set_yticks([])\n",
    "    axes[-1].set_ylabel(cat1, fontsize=16, color='black', alpha=1)\n",
    "    axes[-1].yaxis.set_label_coords(-0.15,0.5)\n",
    "    \n",
    "    patches = [[]] * (len(axes)-1) \n",
    "    counter=0\n",
    "    for i in axes:\n",
    "        if counter > (len(patches) -1):\n",
    "            break\n",
    "        current_handles, current_labels = axes[counter].get_legend_handles_labels()\n",
    "        patches[counter] = ((current_handles[0].get_edgecolor()).tolist()[0])\n",
    "        counter+=1\n",
    "    \n",
    "    \n",
    "    counter = 0\n",
    "    labels = list(adata.obs[cat1].unique())\n",
    "    lengend_patches = []\n",
    "    \n",
    "    gmeans = []\n",
    "    for i in patches:\n",
    "        \n",
    "        if scale_log:\n",
    "            std_ = round(data[data[cat1].isin([labels[counter]])]['log_scaled_col'].std())\n",
    "            kur_ = round(stats.kurtosis(data[data[cat1].isin([labels[counter]])]['log_scaled_col']), 2)\n",
    "            gmean_ = round(stats.gmean(data[data[cat1].isin([labels[counter]])]['log_scaled_col']))\n",
    "            gmeans.append(gmean_)\n",
    "        else:\n",
    "            std_ = round(data[data[cat1].isin([labels[counter]])][cat2].std())\n",
    "            kur_ = round(stats.kurtosis(data[data[cat1].isin([labels[counter]])][cat2]), 2)\n",
    "            gmean_ = round(stats.gmean(data[data[cat1].isin([labels[counter]])][cat2]))\n",
    "            gmeans.append(gmean_)\n",
    "        lengend_patches.append(mpatches.Patch(color=i, label=r'{a}:     {b},     {c},     {d}'.format(a=labels[counter], b=std_, c=kur_, d=gmean_))) #'\n",
    "        counter+=1\n",
    "    \n",
    "        \n",
    "    legend1 = plt.legend(handles=lengend_patches, title='Section:         std,        K,        GM', bbox_to_anchor=(1.6, 1), labelspacing = 2, fontsize=12, title_fontsize=15)\n",
    "    #plt.setp(legend.get_title(), color='red')\n",
    "    \n",
    "    counter=0\n",
    "    for i in patches:\n",
    "        #if scale_log == True:\n",
    "        #    X = np.log10(gmeans[counter])\n",
    "        #else:\n",
    "        #    X = gmeans[counter]\n",
    "        X = gmeans[counter]\n",
    "        axes[counter].axvline(X, color='red', lw=2, alpha=1, ymax=0.1)\n",
    "        \n",
    "        #if counter == (len(patches)-1):\n",
    "        #    plt.text(x=gmeans[counter] - (gmeans[counter])/4, y=-0.05, s='Geometric mean', alpha=1, fontdict = {'color': 'r', 'fontsize' : '14'}) #axes[counter]\n",
    "            \n",
    "        counter+=1\n",
    "    \n",
    "    \n",
    "    # 2nd legend\n",
    "    \n",
    "    line1 = Line2D([], [], color='red', marker='|', linestyle='None',\n",
    "                              markersize=10, markeredgewidth=1.5, label='Geometric mean')\n",
    "    \n",
    "    legend2 = plt.legend(handles=[line1], title='Graphical overlays', bbox_to_anchor=(1.41, 0.25), labelspacing = 2, fontsize=12, title_fontsize=15)\n",
    "    \n",
    "    plt.gca().add_artist(legend1)\n",
    "    \n",
    "\n",
    "######################################################################################\n",
    "\n",
    "else:\n",
    "\n",
    "    if mode == 'manual':\n",
    "        values = manual_values\n",
    "    \n",
    "    elif mode == 'gene_expression':\n",
    "        if len(plot_value) > 1:\n",
    "            raise Exception(\"Current implementation only supporting single gene expression\")\n",
    "        \n",
    "        \n",
    "        df_of_values = (adata.varm[adata.uns['masks'][mask_set]['varm']].T)[plot_value]\n",
    "        df_of_values = df_of_values.T\n",
    "        values = []\n",
    "        for col in df_of_values:\n",
    "            value = list(df_of_values[col].values)\n",
    "            values.extend(value)\n",
    "            \n",
    "    elif mode in ['celltype_counts','celltype_percentage_within_sections','celltype_percentage_across_sections']:\n",
    "    \n",
    "        \n",
    "        if len(plot_value) > 1:\n",
    "            adata.obs['combined_annotation'] = adata.obs[cat2].copy().astype(str)\n",
    "            for value in plot_value:\n",
    "                adata.obs.loc[adata.obs[cat2].isin([value]), 'combined_annotation'] = 'combined_annotation'\n",
    "            cat2 = 'combined_annotation'\n",
    "            plot_value = ['combined_annotation']\n",
    "        \n",
    "        adata.obs[cat1] = adata.obs[cat1].astype('category')\n",
    "        adata.obs[cat2] = adata.obs[cat2].astype('category')\n",
    "        \n",
    "        # generate counts table\n",
    "        counts_table = pd.crosstab(adata.obs[cat1],adata.obs[cat2])\n",
    "        \n",
    "        if mode == 'celltype_counts':\n",
    "            df_of_values = counts_table[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "        \n",
    "        elif mode == 'celltype_percentage_across_sections':\n",
    "            percentage_table_column = (round((counts_table/counts_table.sum())*100,2))\n",
    "            df_of_values = percentage_table_column[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "                \n",
    "                \n",
    "        elif mode == 'celltype_percentage_within_sections':\n",
    "            percentage_table_row = (round(((counts_table.T)/(counts_table.T).sum())*100,2)).T\n",
    "            df_of_values = percentage_table_row[plot_value]\n",
    "            values = []\n",
    "            for col in df_of_values:\n",
    "                value = list(df_of_values[col].values)\n",
    "                values.extend(value)\n",
    "        \n",
    "    else:\n",
    "        raise Exception('Mode option not correct. Please use one of the following: manual, celltype_counts, celltype_percentage or gene_expression')\n",
    "        \n",
    "######################################################################################\n",
    "        \n",
    "    # create a color scale on the range of values inputted\n",
    "    \n",
    "    values_dict = dict(zip(adata.obs[cat1].unique(),values))\n",
    "    if scale == 'auto':\n",
    "        norm = mpl.colors.Normalize( vmin=min(values) , vmax=max(values) )\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "        \n",
    "    elif scale == 'manual':\n",
    "        norm = mpl.colors.Normalize( vmin=scale_lower_value , vmax=scale_upper_value )\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    \n",
    "    else:\n",
    "        raise Exception('Scale option not correct. Please use either auto or manual')    \n",
    "        \n",
    "######################################################################################\n",
    "    \n",
    "    base_img = np.full(adata.uns[mask_set+'_shape'], 255, dtype=np.uint8)\n",
    "    \n",
    "    for n in adata.uns[mask_set+'_polygons'].keys():\n",
    "        cv2.fillPoly(base_img, pts=tuple([adata.uns[mask_set+'_polygons'][n][0]]), color=tuple(list(int((255*x)) for x in list(sm.to_rgba([val for key, val in values_dict.items() if key in n][0]))[0:3])))\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "######################################################################################\n",
    "    \n",
    "        \n",
    "    # plot the embryo which can be used in all modes for categorical plotting\n",
    "    plt.figure(figsize=(20,16))\n",
    "    \n",
    "    plt.imshow(base_img)\n",
    "    \n",
    "    plt.axis('off')\n",
    "    \n",
    "######################################################################################\n",
    "    \n",
    "    if len(plot_value) == len(adata.obs[cat2].unique()):\n",
    "        plt.title(f'All options in category selected', fontsize=40, y=1.05)\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        # add scale bar to plot\n",
    "        cb = plt.colorbar(sm)\n",
    "        cb.ax.tick_params(labelsize=20)\n",
    "    \n",
    "    \n",
    "        if mode == 'manual':\n",
    "            plt.title('Manual values for every section', fontsize=40, y=1.05)\n",
    "            cb.set_label(\"Manual values\", fontsize=30, rotation=270, labelpad=70)\n",
    "            \n",
    "        elif mode == 'gene_expression':\n",
    "            plt.title(f'Mean gene expression of {plot_value[0]} for each section',fontsize=40, y=1.05)\n",
    "            \n",
    "            if scale == 'manual':\n",
    "                cb.set_label(\"Expression\" + \" (scale values representitive to manual set upper and lower thresholds)\", fontsize=20, rotation=270, labelpad=70)\n",
    "            elif scale == 'auto':\n",
    "                cb.set_label(\"Expression\", fontsize=30, rotation=270, labelpad=70)\n",
    "            \n",
    "        elif mode in ['celltype_counts','celltype_percentage_within_sections','celltype_percentage_across_sections']:\n",
    "              \n",
    "            if mode == 'celltype_counts':\n",
    "                plt.title(f'Number of counts for {plot_value[0]}', fontsize=40, y=1.05)\n",
    "            \n",
    "                if scale == 'manual':\n",
    "                    cb.set_label(\"No. cells\" + \" (scale values representitive to manual set upper and lower thresholds)\", fontsize=20, rotation=270, labelpad=70)\n",
    "                elif scale == 'auto':\n",
    "                    cb.set_label(\"No. cells\", fontsize=30, rotation=270, labelpad=70)\n",
    "                \n",
    "            elif mode == 'celltype_percentage_within_sections':\n",
    "                plt.title(f'Percentage of {plot_value[0]} compared within section', fontsize=40, y=1.05)\n",
    "                \n",
    "                if scale == 'manual':\n",
    "                    cb.set_label(\"Percentage %\" + \" (scale values representitive to manual set upper and lower thresholds)\", fontsize=20, rotation=270, labelpad=70)\n",
    "                elif scale == 'auto':\n",
    "                    cb.set_label(\"Percentage %\", fontsize=30, rotation=270, labelpad=70)\n",
    "            \n",
    "            elif mode == 'celltype_percentage_across_sections':\n",
    "                plt.title(f'Percentage of {plot_value[0]} across sections', fontsize=40, y=1.05)\n",
    "                \n",
    "                if scale == 'manual':\n",
    "                    cb.set_label(\"Percentage %\" + \" (scale values representitive to manual set upper and lower thresholds)\", fontsize=20, rotation=270, labelpad=70)\n",
    "                elif scale == 'auto':\n",
    "                    cb.set_label(\"Percentage %\", fontsize=30, rotation=270, labelpad=70)\n",
    "    \n",
    "######################################################################################\n",
    "\n",
    "if save_plot != False:\n",
    "    plt.savefig(save_plot)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python_env_Hlab",
   "language": "python",
   "name": "python_env_hlab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
